Moller-Plesset/DF-MP2_NAF.py

```diff
@@ -0,0 +1,210 @@
+Instead cheap intermediates are used the reduced Qov tensor is regained as the last step.
+
+References: 
+1. Algorithm modified from Rob Parrish's most excellent Psi4 plugin example
+Bottom of the page: http://www.psicode.org/developers.php
+2. Tutorials/03_Hartree-Fock/density-fitting.ipynb
+3. M. Kállay, J. Chem. Phys. 2014, 141, 244113. [http://aip.scitation.org/doi/10.1063/1.4905005]
+"""
+
+__authors__ = "Holger Kruse"
+__credits__ = ["Daniel G. A. Smith", "Dominic A. Sirianni"]
+
+__copyright__ = "(c) 2014-2018, The Psi4NumPy Developers"
+__license__ = "BSD-3-Clause"
+__date__ = "2018-11-29"
+
+import time
+import numpy as np
+np.set_printoptions(precision=5, linewidth=200, suppress=True)
+import psi4
+
+# Set memory & output
+psi4.set_memory('2 GB')
+psi4.core.set_output_file('output.dat', False)
+
+mol = psi4.geometry(""" 
+C    1.39410    0.00000   0.00000
+C    0.69705   -1.20732   0.00000
+C   -0.69705   -1.20732   0.00000
+C   -1.39410    0.00000   0.00000
+C   -0.69705    1.20732   0.00000
+C    0.69705    1.20732   0.00000
+H    2.47618    0.00000   0.00000
+H    1.23809   -2.14444   0.00000
+H   -1.23809   -2.14444   0.00000
+H   -2.47618    0.00000   0.00000
+H   -1.23809    2.14444   0.00000
+H    1.23809    2.14444   0.00000
+symmetry c1
+""")
+
+# Basis used in mp2 density fitting
+psi4.set_options({'basis': 'aug-cc-pVDZ', 'df_basis_mp2': 'aug-cc-pvdz-ri'})
+
+check_energy = True
+
+print('\nStarting RHF...')
+t = time.time()
+RHF_E, wfn = psi4.energy('SCF', return_wfn=True)
+print('...RHF finished in %.3f seconds:   %16.10f' % (time.time() - t, RHF_E))
+
+# Grab data from Wavfunction clas
+ndocc = wfn.nalpha()
+orbital_basis = wfn.basisset()
+nbf = wfn.nso()
+nvirt = nbf - ndocc
+
+# Split eigenvectors and eigenvalues into o and v
+eps_occ = np.asarray(wfn.epsilon_a_subset("AO", "ACTIVE_OCC"))
+eps_vir = np.asarray(wfn.epsilon_a_subset("AO", "ACTIVE_VIR"))
+
+# Build DF tensors
+print('\nBuilding DF ERI tensor Qov...')
+t = time.time()
+C = wfn.Ca()
+
+# Build instance of MintsHelper
+mints = psi4.core.MintsHelper(orbital_basis)
+zero_bas = psi4.core.BasisSet.zero_ao_basis_set()
+
+# build auxiliary basis set object
+aux_basis = psi4.core.BasisSet.build(
+    mol, "DF_BASIS_MP2", "", "RIFIT",
+    psi4.core.get_global_option('df_basis_mp2'))
+naux = aux_basis.nbf()
+
+# Build (P|pq) raw 3-index ERIs, dimension (1, Naux, nbf, nbf)
+Ppq = mints.ao_eri(zero_bas, aux_basis, orbital_basis, orbital_basis)
+
+# Build Coulomb metric but only invert, dimension (1, Naux, 1, Naux)
+metric = mints.ao_eri(zero_bas, aux_basis, zero_bas, aux_basis)
+metric.power(-1.0, 1.e-14)
+
+# Remove excess dimensions of Ppq, & metric
+Ppq = np.squeeze(Ppq)
+metric = np.squeeze(metric)
+
+# paper uses transpose of Ppq, so we adapt for now
+Ppq = np.reshape(Ppq, (naux, nbf * nbf)).T
+print("I  = (pq|P) dim:", Ppq.shape)
+
+# cholesky decomp of inverse metric
```

Comment:
Could you add a few links back the papers equations on a few of these comments here?